jsonschema/reference.go

```diff
@@ -0,0 +1,161 @@
+package jsonschema
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/franela/goreq"
+)
+
+// jsonRef represents a JSON Reference source and targetRef
+type jsonRef struct {
+	Source []string
+	Target string
+}
+
+// dereference parses JSON string and replaces all $ref with the referenced data.
+func dereference(schemaPath string, input []byte, isTop bool) ([]byte, error) {
+	if !strings.Contains(string(input), "$ref") {
+		return input, nil
+	}
+
+	var data interface{}
+	json.Unmarshal([]byte(input), &data)
+	refs, err := walkInterface(data, []string{}, []jsonRef{})
+	if err != nil {
+		return input, fmt.Errorf("unable to walk interface %s: %v", schemaPath, err)
+	}
+
+	for _, ref := range refs {
+		top := data
+		for i, item := range ref.Source {
+			if isTop && (ref.Source[0] == "allOf" || ref.Source[0] == "oneOf") {
+				continue // do not dereference top-level allOr and oneOf
+			}
+			if i < len(ref.Source)-1 { // iterate
+				if intKey, err := strconv.Atoi(item); err == nil {
+					top = top.([]interface{})[intKey]
+				} else {
+					top = top.(map[string]interface{})[item]
+				}
+			} else { // set reference
+				targetRef, err := buildReference(schemaPath, data, ref.Target)
+				if err != nil {
+					return input, fmt.Errorf("unable to build reference from %s: %v", ref.Target, err)
+				}
+				targetKeys := reflect.ValueOf(targetRef).MapKeys()
+				if len(targetKeys) > 1 {
+					// assuming integer item is slice[index] instead of map[string]
+					if intKey, err := strconv.Atoi(item); err == nil {
+						top.([]interface{})[intKey] = targetRef
+					} else {
+						top.(map[string]interface{})[item] = targetRef
+					}
+				} else {
+					// when targetRef = single KV pair, set the value using the key instead of overwriting entire map
+					key := targetKeys[0].Interface().(string)
+					// assuming integer item is slice[index] instead of map[string]
+					if intKey, err := strconv.Atoi(item); err == nil {
+						top.([]interface{})[intKey].(map[string]interface{})[key] = targetRef.(map[string]interface{})[key]
+						delete(top.([]interface{})[intKey].(map[string]interface{}), "$ref")
+					} else {
+						top.(map[string]interface{})[item].(map[string]interface{})[key] = targetRef.(map[string]interface{})[key]
+						delete(top.(map[string]interface{})[item].(map[string]interface{}), "$ref")
+					}
+				}
+			}
+		}
+	}
+
+	return json.Marshal(data)
+}
+
+// walkInterface traverses the map[string]interface{} to located json references
+func walkInterface(node interface{}, source []string, refs []jsonRef) ([]jsonRef, error) {
+	var err error
+	for key, val := range node.(map[string]interface{}) {
```

Comment:
Another unchecked type assertion here, like I said below you should either check it or just make node be a map[string]interface{} in the function arguments.

Reply:
I don't think I can make node a map[string]interface{} due to it changing between string/slice/map during recursion, so will add in the assertion check

Reply:
Was not able to add the assertion check INSIDE parseReference due to the multi/single-level conflict, so I added before the parseReference call.

Reply:
I'm not quite following, can't you just make this
`
nodeMap, ok := node.(map[string]interface{})
if ! ok {
  return nil //or whatever is right}
for key, val := range nodeMap {
`

Reply:
That syntax creates this error: `unexpected if at end of statement` --- so I fixed it up by adding the necessary semi-colons: `nodeMap, ok := node.(map[string]interface{}); if ! ok { return nil //or whatever is right}; for key, val := range nodeMap {` but then my Atom will auto-format and remove the semicolons and put them on separate lines

Reply:
my bad I forgot to tripple backtic the comment, separate lines is what I intended.